### ISSUE TITLE
Mounted twice when layout change

### ISSUE THREAD

**Issue by usb248** [nuxt/nuxt.js#5703]
### Version

[v2.6.2](https://github.com/nuxt.js/releases/tag/v2.6.2)

### Reproduction link

[https://codesandbox.io/s/wn7j570y85](https://codesandbox.io/s/wn7j570y85)

### Steps to reproduce

Simply click on button named "Mounted triggered twice, see console" and see in console.
(reproducable only with css:false for transition)

### What is expected ?

mounted once

### What is actually happening?

mounted twice

### Additional comments?

The bug is still present on last release 2.7.1
It can be fixed quickly ? it's really a brake on client-side performance.
(Related : https://github.com/nuxt/nuxt.js/issues/5616)

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c9188">#c9188</a>)</em></sub></div>

**Comment by Austio** [comment]
![image](https://user-images.githubusercontent.com/1709860/57585670-cf26bf80-74b0-11e9-8b5b-c930e70b3cb7.png)

I pulled this down locally and am only seeing a single mount.

**Comment by usb248** [comment]
no, you forgot something. The next page must have an other layout and css:false for transition. I have the same bug locally.
I have to make a video to prove it to you?

**Comment by usb248** [comment]
Someone is alive ? :)

**Comment by ricardogobbosouza** [comment]
Hi @usb248 
This still happens in version `2.8.0`
The problem is when `transition: {css: false}` it's defined

**Comment by usb248** [comment]
Any core team feels concerned ?

**Comment by byshine** [comment]
It's not just transition, still mounts twice on  `this.$router.push`

**Comment by amritk** [comment]
This seems to still be an issue, event with no transitions and two bare bones layouts, when you switch layouts the page gets double created/mounted

**Comment by amritk** [comment]
This still seems to be an issue when transitioning from one layout to another, could we get a re-open?

**Comment by swapneelvivekdesaithoughtworks** [comment]
Happening with Nuxt version 2.10.0 as well. I am using nuxt generate. It only happens in the production build.

**Comment by amritk** [comment]
https://codesandbox.io/s/nuxt-vuetify-k7yv9 just hit the link and watch the console, you see created twitce

**Comment by amritk** [comment]
Hit the link and watch the console
https://codesandbox.io/s/nuxt-vuetify-k7yv9
![2019-10-08:11:54:31](https://user-images.githubusercontent.com/2039539/66424469-8bce6f80-e9c2-11e9-9e11-3b6e3a16ea38.png)

**Comment by amritk** [comment]
It is still an issue:

https://codesandbox.io/s/nuxt-vuetify-k7yv9

**Comment by marcbelletre** [comment]
I can confirm this does still happen with Nuxt 2.10.2 when page transitions are enabled.  

I used to have these objects in my Nuxt config file in order to enable transitions:
```
layoutTransition: {
    name: 'layout',
    mode: ''
},

pageTransition: {
    name: 'page',
    mode: ''
}
```

I had to remove both of them to prevent pages components to be mounted/created twice.

**Comment by MohitTilwani15** [comment]
As a temporary fix, I wrapped my code inside `mounted` in `if` condition. The condition is just to check if the layout name is the one I am expecting.

```
if (this.$nuxt.layoutName === 'myLayout') {
  ...
}
```

**Comment by mrdj07** [comment]
Don't know if this can help, but I managed to find my problem by putting breakpoints inside the fetch methode to realize the page was being loaded twice, by two different methods. 

We had an el-menu made by Element-UI, that had inside of it, nuxt-links, so the route was loaded twice simultaneously. 

I realize this was a coding error but this might help anyone looking for awnsers. Try to trace how the route gets changed !

**Comment by dsktschy** [comment]
My temporary workaround.

~/plugins/inject-layout.js
```js
export default (context, inject) => {
  inject('layout', {

    // Layout name and page path of pages that is not dafault layout
    map: {
      'one-column': ['/signup/', '/login/'],
      'some-other-layout': ['/some/page/path/']
    },

    // Property that represents correct layout name of current page
    current: 'default'
  })
}
```

nuxt.config.js
```js
export default {
  plugins: ['~/plugins/inject-layout.js']
}
```

~/mixins/layout.js
```js
export default {

  // Set and keep correct layout name of current page
  layout({ app, route }) {
    let layoutName = 'default'
    for (const [key, value] of Object.entries(app.$layout.map)) {
      if (value.includes(route.path)) layoutName = key
    }
    app.$layout.current = layoutName
    return layoutName
  },
  
  computed: {
    // True while duplicated lifecycle hooks are running
    incorrectLayout() {
      return this.$nuxt.layoutName !== this.$layout.current
    }
  }
}
```

~/pages/some-page.vue
```js
import layout from '~/mixins/layout'

export default {

  // Layout setting
  mixins: [layout],

  // Hooks will be executed twice
  mounted() {
    // Return immediately if duplicated runs
    if (this.incorrectLayout) return
    console.log('This is only displayed once.')
  }
}
```

**Comment by galaxyblur** [comment]
This is still occurring for me as of today, in nuxt@2.11.0. My Vue browser extension reports that in the default layout, there are two <nuxt /> views and thus everything occurs twice. This is in dev mode.

While running in dev mode, if I remove the <nuxt /> element from layouts/default.vue, the page structure is cleared. Adding it back, one <nuxt /> view appears (according to the Vue browser extension). After that, refreshing the page results in two <nuxt /> views, indefinitely.

Given that this is happening within Nuxt, I don't know of any workarounds or logical statements that can prevent it.

<img width="417" alt="image" src="https://user-images.githubusercontent.com/4163126/74049127-7e8db580-49a1-11ea-855e-58ffa1c405a8.png">

**Comment by gerasimvol** [comment]
similar  to https://github.com/nuxt/nuxt.js/issues/6961

**Comment by nickgayo12** [comment]
I have the same issue with twice mounted on nuxt, what I did is this.

`<client-only>
    <nuxt /> // On your layout
<client-only>`

**Comment by tsarrB** [comment]
I had something similar
the problem was in the incorrectly registered plugin

**Comment by gerasimvol** [comment]
@tsarrB can you provide any details? examples?

**Comment by tsarrB** [comment]
@gerasimvol 
my case is described [here](https://medium.com/@valerka/how-to-register-the-plugin-correctly-in-nuxt-js-91ed2d95b81f)

**Comment by 13VF** [comment]
That's still a huge problem, some components mounts twice - that's lead to visual bugs (for example, i have component that mount maps-script, and that leads for two maps visible).

I disabled SSR, cause site now in just demo mode, but I need SSR soon.

Any updates?

**Comment by crzyzenn** [comment]
I'm facing the same issue! Using "nuxt": "^2.11.0". Need to check layout name before firing stuff inside created hook to prevent firing twice :(

**Comment by LuisPao** [comment]
Same issue. `beforeDestroy` fires twice too，and  I cannot get the right `this.$nuxt.layoutName` in beforeDestroy

**Comment by bbarbosa85** [comment]
Found another "fix" using Vuex / Nuxt Store. 

By putting a 'v-if' statement on the <nuxt/> element on the layout itself you can prevent it from loading the content of the next route before the layout changes. 

This way you won't need to deal with every 'created' or 'mounted' function on your page components.

What I did:

**Step 1)** create a ```pages.js``` store module to register the content status.
IMPORTANT: to use vuex modules like this, you need to create an ```index.js``` file on ```~/store``` folder
> ~/store/pages.js
```javascript
export const state = () => ({
  showLayoutContent: true // show by default in case of first access or refreshing the page
})

export const mutations = {
  TOGGLE_LAYOUT_CONTENT (state, isVisible) {
     state.showLayoutContent = isVisible
  },
}
```

**Step 2)** Create an ```app.js``` middleware to hide the content on every route change
> ~/middleware/app.js
```javascript
export default function (context) {
  // hide the content
  context.store.commit('pages/TOGGLE_LAYOUT_CONTENT', false)
}
```

**Step 3** Add the middleware reference to your ```nuxt.config.js``` to extend the router property
> ~/nuxt.config.js
```javascript
//{...}
router: {
    middleware: ['app']
},
//{...}
```

**Step 4** Enable and use the store on your layout. 
```

In my case I've used the ```default.vue``` layout as a base of extension for my other layouts so the computed method is created

**Comment by gerasimvol** [comment]
same here https://github.com/nuxt/nuxt.js/issues/6961
cause serious visual issues 🙁

**Comment by Atinux** [comment]
Will look at it, sorry for the delay 😬

**Comment by Atinux** [comment]
Update: the issue seems to come only when doing server-side rendering (no issue with `nuxt dev --spa`).

I looked at the Vue Devtools and indeed, the default layout looked different when server-rendered vs client-side (when navigating on error page for example)

SSR:
<img width="356" alt="Screenshot 2020-03-25 at 17 57 37" src="https://user-images.githubusercontent.com/904724/77563878-26176680-6ec2-11ea-8b18-33d6f1a6dc71.png">

After navigating:
<img width="303" alt="Screenshot 2020-03-25 at 17 57 41" src="https://user-images.githubusercontent.com/904724/77563900-2ca5de00-6ec2-11ea-9e29-fc6a62cd8558.png">

So it actually re-render the whole layout, causing the new mounted() hooks.

Keeping the investigation 🕵🏻‍♂️

**Comment by Atinux** [comment]
I found the issue, apparently the `name` of the layout is not defined, which cause a different name between server-side rendering and client-side (`Default` on SSR and `layouts/default.vue` on client).

**As a quick workaround:** Set a name for each of your layouts:

`layouts/default.vue`
```vue
<template>...</template>

<script>
export default {
  name: 'default'
}
</script>
```

I will make a PR to fix it in the next patch release of NuxtJS.

Thank you very much for reporting this and sorry for this outstanding delay 😞

**Comment by Atinux** [comment]
Fixed in V2.12.1 release 🔥

**Comment by asolopovas** [comment]
For me, it is still calls mounted twice.

**Comment by pi0** [comment]
@asolopovas Would you please fill in a new issue with reproduction code?

**Comment by Chheung** [comment]
Still happens in v2.12.2

**Comment by manniL** [comment]
> Still happens in v2.12.2

Then create a new issue and a reproduction CodeSandbox or repo 😉

**Comment by falstack** [comment]
@manniL why not reopen this issue? i have same issue

**Comment by falstack** [comment]
```html
<template>
  <div id="layout-web">
    <span> <!--  there has a span，it works -->
      <nuxt />
    </span>
    <SignDialog />
  </div>
</template>
```

```html
<template>
  <div id="layout-web">
    <nuxt /> <!--  has same level html, has bug, mounted twice  -->
    <SignDialog />
  </div>
</template>
```

**Comment by mhrabiee** [comment]
Same issue on 2.12.2

**Comment by c0bra** [comment]
Yep, still happening on 2.12.2

**Comment by falstack** [comment]
still happening 

![image](https://user-images.githubusercontent.com/16357724/82321515-8846a080-9a07-11ea-8645-0af3a682762a.png)

**Comment by manniL** [comment]
> Then create a new issue and a reproduction CodeSandbox or repo 😉

**Comment by nhuthuynh** [comment]
@Atinux  
First of all, thanks for you effort and suggestion!

Using nuxt  2.12.2 and still have the problem
I am using both SPA and SSR

![Screen Shot 2020-05-21 at 11 51 32](https://user-images.githubusercontent.com/3124346/82514598-812fa780-9b59-11ea-9293-7dd96babde5e.png)

![Screen Shot 2020-05-21 at 11 54 22](https://user-images.githubusercontent.com/3124346/82514801-fd29ef80-9b59-11ea-8ddc-c5fd6c6ee017.png)

<img width="433" alt="Screen Shot 2020-05-21 at 11 52 35" src="https://user-images.githubusercontent.com/3124346/82514797-fa2eff00-9b59-11ea-8f8a-f1a84ec77e2c.png">

> I found the issue, apparently the `name` of the layout is not defined, which cause a different name between server-side rendering and client-side (`Default` on SSR and `layouts/default.vue` on client).
> 
> **As a quick workaround:** Set a name for each of your layouts:
> 
> `layouts/default.vue`
> 
> ```
> <template>...</template>
> 
> <script>
> export default {
>   name: 'default'
> }
> </script>
> ```
> 
> I will make a PR to fix it in the next patch release of NuxtJS.
> 
> Thank you very much for reporting this and sorry for this outstanding delay 😞

tried this but have not resolved, any other suggestion :(

**Comment by simplenotezy** [comment]
Hi - I am getting this error as well. I followed this pattern to create nested layout: https://constantsolutions.dk/2020/02/nested-layouts-in-nuxt-vue-js/ I am not sure if the other guys uses "slots" as well in their layout, and that's why it is happening?

**Comment by simplenotezy** [comment]
As an example, in my pages/_slug.vue I have this:

    mounted () {
        console.log('Mounted!', this.$nuxt.layoutName, this.$nuxt);
    }

And in the console I get:

![image](https://user-images.githubusercontent.com/3786627/83244073-23c1d900-a19f-11ea-8ffc-54ae59c224d7.png)

The strange thing is that no matter what I set in the layout name, it will always return "page" or "default". Changing page layout name to "name: 'APage'," it will still say it mounted "page".

I believe this could be the reason why my Nuxt application is slow when changing pages, despite having an API response time of 50ms.

**Comment by falstack** [comment]
reopen this issue?

**Comment by simplenotezy** [comment]
Hi @manniL, I wanted to make a reproducible case, so I forked @amritk's codesandbox, and updated to Nuxt version **2.12.2** - I can confirm I didn't have to make any changes for the error to appear.

It writes "I have been mounted" twice in the console, see here: https://codesandbox.io/s/nuxt-vuetify-0xz91?file=/package.json. I even tried to give a name to the layout file.

When switching to the about page:

```
i have been created. Layout: default 
i have been mounted. Layout: default 
i have been created. Layout: other 
i have been mounted. Layout: other 
```

It's also weird it calls the layout "other" when I call `this.$nuxt.layoutName`, because I explicitly set the name to be `otherlayout`.

I think it would be appropriate to reopen this issue. Also I believe it causes page switches to take extra long time - we just launched our production site: https://thejewelleryroom.com and when switching pages to a page that has been cached by Cloudflare, and the response time is 50-100ms, it still takes 1-2 seconds for the page to switch. Perhaps it is because it mounts the page twice?

**Comment by melihplt** [comment]
This is a very serious bug. I cannot not generate unique id for a client side only map container because it is calculated twice. I can not push a value to my empty-by-default array because it is pushed twice.

**Comment by simplenotezy** [comment]
Issue reopened here: https://github.com/nuxt/nuxt.js/issues/7441

**Comment by simplenotezy** [comment]
@melihplt you can make this workaround in your mounted/created hook.

    if (this.$nuxt.layoutName === 'myLayout') {
        // will only fire once
    }

However, it does not solve the performance issues it generates.

**Comment by Solero007** [comment]
Any ideas? Having the same issue

**Comment by safiullahsarhandi** [comment]
> I have the same issue with twice mounted on nuxt, what I did is this.
> 
> `<client-only> <nuxt /> // On your layout <client-only>`
> 
> hope it helps.

this works for me

**Comment by stouch** [comment]
This helped me :

> Solution: After plucking some hair, I found that simply wrapping `<Nuxt />` in any element fixed the issue. No idea why, but it works.
> 
> ```
> <template>
>   <div>
>     <TheHeader />
>     <div>
>         <Nuxt />
>     <div>
>     <TheFooter />
>   </div>
> </template>
> ```

But I don't understand why this ticket is "closed" ? The problem is still happening in 2.15.8 .

**Comment by simeon-petrov-5** [comment]
I can also confirm this as of 2.15.7 SSR Mode (from the previous comment 2.15.8 isn't any different). What helped was wrapping `<Nuxt />` in a `div` (I also have a header and a footer inside my layout). 

I have page transitions still enabled (`transition: 'page-transition'`) so I guess it's not only tied to the transitions.

> Solution: After plucking some hair, I found that simply wrapping `<Nuxt />` in any element fixed the issue. No idea why, but it works.
> 
> ```
> <template>
>   <div>
>     <TheHeader />
>     <div>
>         <Nuxt />
>     <div>
>     <TheFooter />
>   </div>
> </template>
> ```

**Comment by kamaludin21** [comment]
Same issue here

**Comment by trandaison** [comment]
Why close this issue? It's still happen on 2.15.7

**Comment by trandaison** [comment]
In my case, I have a component `TheNavbar` in the layout file.

```vue
<template>
  <div>
    <div class="default-layout">
      <slot>
        <Nuxt />
      </slot>
    </div>
    <TheNavbar />
  </div>
</template>
```

In this component, I have a computed that mapped from vuex store. And the navbar will render links dynamically based on that computed.

I have to put that template code that uses the vuex state within `<client-only>` component.

The state from my vuex is set when a client side plugin executing. If it was set from the server side this will not happen, I guess.

**Comment by ahmetcankucukkor** [comment]
I am having this problem in version 2.14.3. When I get data from the cookie, the components are mounted twice.

I have a link list in footer. I need to show one of these links when language is 'tr' only. 

```
<li v-if="checkLanguage() === 'tr'">
    <a
        :href="imageURL + 'files/kvkkApply.docx'"
        download
        rel="nofollow">
        KVKK Başvuru Formu
    </a>
</li>
```

In checkLanguage(), I am getting language data from Cookie.

```
checkLanguage () {
      return Cookie.get("i18n_redirected");
}
```

I don't get any errors while using them, but the components are mounted twice.
If I do not use the cookie in the checklanguage function and do it like this, the components do not mount 2 times

```
checkLanguage () {
      return 'tr'    //Cookie.get("i18n_redirected");
}
```

**Comment by brunodeangelis** [comment]
> I am having this problem in version 2.14.3. When I get data from the cookie, the components are mounted twice.
> 
> I have a link list in footer. I need to show one of these links when language is 'tr' only.
> 
> ```
> <li v-if="checkLanguage() === 'tr'">
>     <a
>         :href="imageURL + 'files/kvkkApply.docx'"
>         download
>         rel="nofollow">
>         KVKK Başvuru Formu
>     </a>
> </li>
> ```
> 
> In checkLanguage(), I am getting language data from Cookie.
> 
> ```
> checkLanguage () {
>       return Cookie.get("i18n_redirected");
> }
> ```
> 
> I don't get any errors while using them, but the components are mounted twice. If I do not use the cookie in the checklanguage function and do it like this, the components do not mount 2 times
> 
> ```
> checkLanguage () {
>       return 'tr'    //Cookie.get("i18n_redirected");
> }
> ```

Have you tried using `v-show` instead of `v-if`?

**Comment by ahmetcankucukkor** [comment]
> > I am having this problem in version 2.14.3. When I get data from the cookie, the components are mounted twice.
> > I have a link list in footer. I need to show one of these links when language is 'tr' only.
> > ```
> > <li v-if="checkLanguage() === 'tr'">
> >     <a
> >         :href="imageURL + 'files/kvkkApply.docx'"
> >         download
> >         rel="nofollow">
> >         KVKK Başvuru Formu
> >     </a>
> > </li>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In checkLanguage(), I am getting language data from Cookie.
> > ```
> > checkLanguage () {
> >       return Cookie.get("i18n_redirected");
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't get any errors while using them, but the components are mounted twice. If I do not use the cookie in the checklanguage function and do it like this, the components do not mount 2 times
> > ```
> > checkLanguage () {
> >       return 'tr'    //Cookie.get("i18n_redirected");
> > }
> > ```
> 
> Have you tried using `v-show` instead of `v-if`?

No, actually I solved my problem with computed variable.
But I just tried the above case with v-show. When I use v-show instead of v-if, it does not mount twice. Thanks.

**Comment by koos30** [comment]
Had the same issue! This thread was a huge help. For me the problem was having multiple elements directly in the page templates.

What I had before (triggered double mount):
```
<template>
    <div v-if="condition">First element</div>
    <div v-else>Second element</div>
</template>
```

Fixed (single mount):
```
<template>
    <div>
        <div v-if="condition">First element</div>
        <div v-else>Second element</div>
    </div>
</template>
```

I was under the impression you are allowed to have multiple elements in the page template as long as only 1 rendered. Apparently not 🤷‍♂️  Hope this helps anyone

**Comment by wmcmurray** [comment]
In my case, one page had an `asyncData` that was fetching some data from an API and adding that data to the store. But it WAS NOT RETURNING THAT DATA, the page was grabbing that data from the store instead (through a computed even).

Components stopped being mounted twice after returning that data directly from the `asyncData` method ! 🎉 (without going through the store)

**Comment by ChicK00o** [comment]
none of the above highlighted solution worked for me on nuxt: 2.15.3

**Comment by Kolobok12309** [comment]
And problem with same root, layout and page not sync changed. Page1 with layout1 would render in layout2, and after it in layout 1

Some debug log for this problem. 
`await this.loadLayout(layout)` from `loadLayoutForNextPage` executed after all `afterEach` and page render in old layout, and after tick(Promise.resolve) layout changed to actual and page mounted second times

```
load unresolved layout landing-v3
end load layout landing-v3
request loaded layout landing-v3
stud mounted
end load of setLayoutForNextPage landing-v3
stud mounted
```

**Comment by vanminhquangtri** [comment]
It's late but I hope this would be useful for some one. 
I my case, the page mounted twice due to the **build** property in **nuxt.config.js**. If I remove build, page will only mount once.
Good luck.

**Comment by trandaison** [comment]
@vanminhquangtri what do you have in the `build` property?
Some project have to configure the build extend or transpile modules, can't be removed easily just like that.

**Comment by vanminhquangtri** [comment]
@trandaison usually because of the **splitChunks** property inside the **build** property. Just try to remove the build property to check whether the error is caused by it.

**Comment by Dmytro-Tihunov** [comment]
I faced with the same problem) are there any ideas ? I have build>transpile that i need )

**Comment by qianzhaoy** [comment]
I solved this problem by overriding `context.app.data` in the `plugins`

```javascript
export default function (Vue, context) {
  const originData = context.app.data;

  context.app.data = function () {
    const data = originData.call(this);
    return {
      ...data,
      layout: 'div',
    };
  };
}
```

**Comment by Kolobok12309** [comment]
> I solved this problem by overriding `context.app.data` in the client-side `plugins`
> 
> ```js
> export default function (Vue, context) {
>   const originData = context.app.data;
> 
>   context.app.data = function () {
>     const data = originData.call(this);
>     return {
>       ...data,
>       layout: 'div',
>     };
>   };
> }
> ```

Can you say more about your fix?

**Comment by Anymfah** [comment]
> This helped me :
> 
> > Solution: After plucking some hair, I found that simply wrapping `<Nuxt />` in any element fixed the issue. No idea why, but it works.
> > ```
> > <template>
> >   <div>
> >     <TheHeader />
> >     <div>
> >         <Nuxt />
> >     <div>
> >     <TheFooter />
> >   </div>
> > </template>
> > ```
> 
> But I don't understand why this ticket is "closed" ? The problem is still happening in 2.15.8 .

Thank you so much

**Comment by damyco** [comment]
> 

This was a real head scratcher for us. This workaround worked though, thanks.

**Comment by AlMuz** [comment]
@danielroe , please help
Nuxt 3.12.4, trying to set layout with setPageLayout according on response from api and unfortunately faced double firing of onMount, onBeforeMount. Also tried to sey layout inside the middleware. 
Everywhere I'm getting exact behavior: page loads - request - getting layout - setting it, fires onMount/onBeforeMount/middleware again.

**Comment by Kolobok12309** [comment]
If any with nuxt 2.x has this problem, try nuxt 2.17.4 or upper, it has fixes #10874 and #27484
